### ISSUE TITLE
Semmle internal error with a c++ project

### ISSUE THREAD

**Issue by vchernin** [github/codeql#9929]
I have been facing this semmle error when trying to run codeql in a arch linux image on github actions. It occurs in the "codeql analysis" step.

<details>
<summary>log</summary>

```sh
Run github/codeql-action/analyze@v2
/usr/bin/docker exec  9665afb8380c9957683ba76598c47c04618894f96c4e03c2cb028f1d1bf33c2f sh -c "cat /etc/*release | grep ^ID"
ERROR: ld.so: object '/__t/CodeQL/0.0.0-20220623/x64/codeql/tools/linux64/${LIB}_${PLATFORM}_trace.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
Failed to call git to get current commit. Continuing with data from environment or input: Error: The process '/usr/sbin/git' failed with exit code 128
Error: The process '/usr/sbin/git' failed with exit code 128
    at ExecState._setResult (/__w/_actions/github/codeql-action/v2/node_modules/@actions/exec/lib/toolrunner.js:592:25)
    at ExecState.CheckComplete (/__w/_actions/github/codeql-action/v2/node_modules/@actions/exec/lib/toolrunner.js:575:18)
    at ChildProcess.<anonymous> (/__w/_actions/github/codeql-action/v2/node_modules/@actions/exec/lib/toolrunner.js:469:27)
    at ChildProcess.emit (node:events:390:28)
    at maybeClose (node:internal/child_process:1064:16)
    at Socket.<anonymous> (node:internal/child_process:450:11)
    at Socket.emit (node:events:390:28)
    at Pipe.<anonymous> (node:net:687:12)
/__t/CodeQL/0.0.0-20220623/x64/codeql/codeql version --format=terse
2.10.0
Finalizing cpp
  /__t/CodeQL/0.0.0-20220623/x64/codeql/codeql database finalize --finalize-dataset --threads=2 /__w/_temp/codeql_databases/cpp --ram=5919
  Running TRAP import for CodeQL database at /__w/_temp/codeql_databases/cpp...
  easyeffects-resources.c.c4be057a.trap (easyeffects-resources.c.c4be057a_0.trap.tar.br) for easyeffects.link, 86736: com.google.common.util.concurrent.UncheckedExecutionException: com.semmle.util.exception.CatastrophicError: String too long (4671914 characters)
  com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
  com.google.common.cache.LocalCache.get(LocalCache.java:3953)
  com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3976)
  com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4960)
  com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4966)
  com.semmle.inmemory.trap.pools.CachingEncoder.encodeString(CachingEncoder.java:30)
  com.semmle.inmemory.ast.ColumnType$4.encode(ColumnType.java:177)
  com.semmle.inmemory.trap.ImportTasksProcessor.checkedEncode(ImportTasksProcessor.java:203)
  com.semmle.inmemory.trap.ImportTasksProcessor.access$000(ImportTasksProcessor.java:61)
  com.semmle.inmemory.trap.ImportTasksProcessor$PerThreadWriter.writeTuple(ImportTasksProcessor.java:291)
  com.semmle.inmemory.trap.TRAPReader.writeMetaTuple(TRAPReader.java:662)
  com.semmle.inmemory.trap.TRAPReader.scanTuple(TRAPReader.java:609)
  com.semmle.inmemory.trap.TRAPReader.scanTuplesAndLabels(TRAPReader.java:523)
  com.semmle.inmemory.trap.TRAPReader.importTuples(TRAPReader.java:423)
  com.semmle.inmemory.trap.ImportTasksProcessor.process(ImportTasksProcessor.java:191)
  com.semmle.inmemory.trap.ImportTasksProcessor.lambda$importTrap$1(ImportTasksProcessor.java:151)
  com.semmle.util.concurrent.FutureUtils.lambda$mapAsync_$8(FutureUtils.java:136)
  java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(Unknown Source)
  java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
  java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
  java.base/java.lang.Thread.run(Unknown Source)
  
   ... caused by:
  
  com.semmle.util.exception.CatastrophicError: String too long (4671914 characters)
  com.semmle.inmemory.util.DiskWritableStringStore.add(DiskPool.java:561)
  com.semmle.inmemory.util.SequentialDiskPool.insertBucket(SequentialDiskPool.java:40)
  com.semmle.inmemory.util.DiskPool.getId(DiskPool.java:254)
  com.semmle.inmemory.stringpool.IntensionalGeneration.encodeString(IntensionalGeneration.java:75)
  com.semmle.inmemory.stringpool.StringPool.encodeString(StringPool.java:173)
  com.semmle.inmemory.trap.pools.SynchronizedEncoder.encodeString(SynchronizedEncoder.java:14)
  com.semmle.inmemory.trap.pools.CachingEncoder$1.load(CachingEncoder.java:23)
  com.semmle.inmemory.trap.pools.CachingEncoder$1.load(CachingEncoder.java:20)
  com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
  com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
  com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
  com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
  com.google.common.cache.LocalCache.get(LocalCache.java:3953)
  com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3976)
  com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4960)
  com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4966)
  com.semmle.inmemory.trap.pools.CachingEncoder.encodeString(CachingEncoder.java:30)
  com.semmle.inmemory.ast.ColumnType$4.encode(ColumnType.java:177)
  com.semmle.inmemory.trap.ImportTasksProcessor.checkedEncode(ImportTasksProcessor.java:203)
  com.semmle.inmemory.trap.ImportTasksProcessor.access$000(ImportTasksProcessor.java:61)
  com.semmle.inmemory.trap.ImportTasksProcessor$PerThreadWriter.writeTuple(ImportTasksProcessor.java:291)
  com.semmle.inmemory.trap.TRAPReader.writeMetaTuple(TRAPReader.java:662)
  com.semmle.inmemory.trap.TRAPReader.scanTuple(TRAPReader.java:609)
  com.semmle.inmemory.trap.TRAPReader.scanTuplesAndLabels(TRAPReader.java:523)
  com.semmle.inmemory.trap.TRAPReader.importTuples(TRAPReader.java:423)
  com.semmle.inmemory.trap.ImportTasksProcessor.process(ImportTasksProcessor.java:191)
  com.semmle.inmemory.trap.ImportTasksProcessor.lambda$importTrap$1(ImportTasksProcessor.java:151)
  com.semmle.util.concurrent.FutureUtils.lambda$mapAsync_$8(FutureUtils.java:136)
  java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(Unknown Source)
  java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
  java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
  java.base/java.lang.Thread.run(Unknown Source)
  at   ""\162\146\141\143\145\076\012\000\000\050\165\165\141\171\051""")
  Oops! A fatal internal error occurred.
  com.semmle.util.exception.CatastrophicError: End of file found while looking for '.pop'.
  	at com.semmle.inmemory.trap.TRAPReader.importTuples(TRAPReader.java:435)
  	at com.semmle.inmemory.trap.ImportTasksProcessor.process(ImportTasksProcessor.java:191)
  	at com.semmle.inmemory.trap.ImportTasksProcessor.lambda$importTrap$1(ImportTasksProcessor.java:151)
  	at com.semmle.util.concurrent.FutureUtils.lambda$mapAsync_$8(FutureUtils.java:136)
  	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(Unknown Source)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
  	at java.base/java.lang.Thread.run(Unknown Source)
  A fatal error occurred: Dataset import for /__w/_temp/codeql_databases/cpp/db-cpp failed with code 100.
  Error: The process '/__t/CodeQL/0.0.0-20220623/x64/codeql/codeql' failed with exit code 2
  Error: The process '/__t/CodeQL/0.0.0-20220623/x64/codeql/codeql' failed with exit code 2
      at toolrunnerErrorCatcher (/__w/_actions/github/codeql-action/v2/lib/toolrunner-error-catcher.js:86:19)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async Object.finalizeDatabase (/__w/_actions/github/codeql-action/v2/lib/codeql.js:582:13)
      at async finalizeDatabaseCreation (/__w/_actions/github/codeql-action/v2/lib/analyze.js:108:13)
      at async runFinalize (/__w/_actions/github/codeql-action/v2/lib/analyze.js:249:5)
      at async run (/__w/_actions/github/codeql-action/v2/lib/analyze-action.js:83:9)
      at async runWrapper (/__w/_actions/github/codeql-action/v2/lib/analyze-action.js:221:9)
  fatal: not a git repository (or any parent up to mount point /)
  Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
  Failed to call git to get current commit. Continuing with data from environment or input: Error: The process '/usr/sbin/git' failed with exit code 128
  Error: The process '/usr/sbin/git' failed with exit code 128
      at ExecState._setResult (/__w/_actions/github/codeql-action/v2/node_modules/@actions/exec/lib/toolrunner.js:592:25)
      at ExecState.CheckComplete (/__w/_actions/github/codeql-action/v2/node_modules/@actions/exec/lib/toolrunner.js:575:18)
      at ChildProcess.<anonymous> (/__w/_actions/github/codeql-action/v2/node_modules/@actions/exec/lib/toolrunner.js:469:27)
      at ChildProcess.emit (node:events:390:28)
      at maybeClose (node:internal/child_process:1064:16)
      at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
```

</details>

<details>
<summary>github actions job</summary>

```yaml
 analyze:
    name: Analyze
    runs-on: ubuntu-latest
    permissions:
      security-events: write
    container:
      image: archlinux
    strategy:
      fail-fast: false
      matrix:
        language: [ 'cpp', 'python' ]

    steps:
    - name: Checkout repository
      uses: actions/checkout@v3.0.2

    - name: Install deps
      run: |
        # install general build deps
        pacman -Syu --noconfirm ninja gcc pkgconf python3 python-pip which
        # install easyeffects build deps
        source ./PKGBUILD && pacman -Syu --noconfirm --needed --asdeps "${makedepends[@]}" "${depends[@]}"

    - name: Initialize CodeQL
      uses: github/codeql-action/init@v2
      with:
        languages: ${{ matrix.language }}

    - name: Autobuild
      uses: github/codeql-action/autobuild@v2

    - name: Perform CodeQL Analysis
      uses: github/codeql-action/analyze@v2
```

</details>

This error only occurs in the c++ part of the workflow, the python one does not show this error (despite many `fatal: not a git repository`).

I see virtually the same stack trace (except no git errors) when locally creating the cpp database with the codeql bundle, except in a different build environment. I used flatpak-builder's build environment for easyeffects, comprised of a few audio related dependencies + gnome runtime, not arch linux).

The repository is https://github.com/wwmm/easyeffects. Is there a way of getting more debug information from semmle?

**Comment by hmakholm** [comment]
Thanks for the report.

4671914 characters is a long string indeed -- since none of the source files in the project is even that long, it looks likely you've hit some bug in our C++ extractor.

("Semmle" was the name of what later became CodeQL after GitHub acquired it. The name is still present in a lot of our internal names, but it just amounts to "internal CodeQL error").

> Is there a way of getting more debug information from semmle?

You're in luck -- we just launched a nice checkbox for collecting more debug info from CodeQL runs. :-) https://github.blog/changelog/2022-08-01-debugging-codeql-analysis-in-code-scanning-made-easier-by-obtaining-detailed-logs-and-debugging-artifacts-from-the-codeql-action/

Could you try pressing that button on a failing run and point us to the debug artifact that produces? Thank you.

**Comment by vchernin** [comment]
> You're in luck -- we just launched a nice checkbox for collecting more debug info from CodeQL runs. :-) https://github.blog/changelog/2022-08-01-debugging-codeql-analysis-in-code-scanning-made-easier-by-obtaining-detailed-logs-and-debugging-artifacts-from-the-codeql-action/

Thanks for the reply, that is some timing :)

I tried collecting the logs by rerurnning all jobs in the workflow with debugging on. The (irrelevant) python debug artifact is uploaded but strangely the c++ debug artifact is missing: https://github.com/vchernin/easyeffects/actions/runs/2776389391. 

I tried looking at the logs via "View raw logs" [from here](https://github.com/vchernin/easyeffects/runs/7616823399), which does show codeql logs with more debug messages, but I couldn't spot any new insight.

I also tried using 
```yaml
with: 
  debug: true
```

And that seems to result in logs being shown here: https://github.com/vchernin/easyeffects/runs/7617688891?check_suite_focus=true (as collapsible sections).

**Comment by hmakholm** [comment]
Well, dangit -- the C++ debug artifact is what I'd have liked to have a look at. I'll ask around internally if there's another way to get at the files I need; stay tuned.

**Comment by hmakholm** [comment]
Thanks for your patience in debugging this!

Try leaving the `debug: true` on, and then add this stanza after after the `codeql-action/analyze` step:
```
- name: Upload
  if: always() && matrix.language == 'cpp'
  uses: actions/upload-artifact@v3
  with:
    name: debug-artifacts-cpp
    path: ${{ runner.temp }}/codeql_databases/cpp
```

**Comment by vchernin** [comment]
Thanks, that seems to have worked: https://github.com/vchernin/easyeffects/actions/runs/2778526550. Only 764 MB

**Comment by hmakholm** [comment]
(Yeah, the fallout is rather large when the extraction process crashes at that point.)

Thanks, I now see what's going on. The build process autogenerates a file `[builddir]/data/easyeffects-resources.c` with a string constant whose content is the byte-for-byte octal encoding of a tarball containing just over a megabyte of XML documents. We then try to store the _C source syntax_ for that constant as a string in the CodeQL database -- that is the 4671914-character string that brings down the house.

Admittedly we ought to handle that case more gracefully; I've filed an internal issue with our C/C++ extractor team. Until that gets fixed, however, your best option is to arrange for that file not to be compiled (or for it to contain shorter dummy data) when you're building for CodeQL analysis. I don't know enough Meson to give concrete advice about that, but you can use https://docs.github.com/en/actions/learn-github-actions/environment-variables to set an environment variable in your workflow that the build system will react to.

Unfortunately CodeQL doesn't provide any feature to specifically ignore a file during extraction.

**Comment by vchernin** [comment]
Thanks, `easyeffects-resource` is a compiled set of UI files, which while large seems like a reasonable thing to do for GTK apps at least.

I only saw your suggested `find` step when writing this comment, so I went on an adventure trying different things. GIO (part of glib, GTK dependency), has an option to [preprocess the xml files and strip blanks](https://docs.gtk.org/gio/struct.Resource.htmll), which would have been great if it it didn't somehow break half our CI builds for reasons I can't explain.

 `../data/easyeffects.gresource.xml: Failed to close file descriptor for child process (Operation not permitted).`

I couldn't reproduce that locally so I just added this before initializing codeql. At least it is probably better than touching internal codeql details.
```yaml
    - name: Minify xml
      run: |
        while IFS= read -r line; do
          xmllint --noblanks "$line" --output "$line"
        done < <( find ./data/ui/ -type f -name "*.ui" -print )
```

**Comment by jketema** [comment]
Hi @vchernin We have a fix for the overly long string issue. This fix should land in the next CodeQL release in a few weeks time. I leave it to you and @hmakholm to decide whether to close the issue, or whether you want to further discuss a workaround.

**Comment by vchernin** [comment]
I’ve subscribed to release notifications to https://github.com/github/codeql-cli-binaries (which seems to release in sync with this repo's tags), so I’ll test without the xmllint workaround whenever I see a new release (and report back).

This workaround works well enough in the meantime, so I’ll close this. Thank for the quick help and fix!

**Comment by vchernin** [comment]
Hi, since a new codeql release v2.10.3 was published I reran the workflow with the workaround removed, but the original error is still there. 

https://github.com/vchernin/easyeffects/runs/7903367483?check_suite_focus=true

**Comment by jketema** [comment]
That run still uses CodeQL 2.10.2: https://github.com/vchernin/easyeffects/runs/7903367483?check_suite_focus=true#step:7:48

**Comment by vchernin** [comment]
> That run still uses CodeQL 2.10.2: https://github.com/vchernin/easyeffects/runs/7903367483?check_suite_focus=true#step:7:48

Ah, I missed that. I assumed setting `github/codeql-action/analyze@v2` would always use the latest `v2.x` release available? I guess `github/codeql-action` does not strictly follow the release schedule of codeql itself.

Anyhow I have retriggered it but with github/codeql-action@main, but that still seems to be v2.10.2: https://github.com/vchernin/easyeffects/runs/7903727916?check_suite_focus=true#step:5:12. I guess I'll try again later and hope for v2.10.3.

**Comment by jketema** [comment]
> Ah, I missed that. I assumed setting github/codeql-action/analyze@v2 would always use the latest v2.x release available? I guess github/codeql-action does not strictly follow the release schedule of codeql itself.

I think the version number in the action has not been bumped yet (it's one of the last things that happens in the release process; after the release of the new CLI)

**Comment by vchernin** [comment]
Hi, this indeed seems to have been fixed since 3.10.4/3.10.3. See https://github.com/wwmm/easyeffects/actions/runs/3064131000/jobs/4946948182#step:7:49.

**Comment by jketema** [comment]
Excellent. Thanks for letting us know.
